Add "Making your dynamic Ember components work with Embroider" blog post (#1706)
simplabs/simplabs.github.io: _posts/2022-02-04-dynamic-components-embroider.md

---
title: 'Making your dynamic Ember components work with Embroider'
author: 'Nick Schot'
github: nickschot
twitter: nickschot
topic: javascript
bio: 'Senior Software Engineer'
og:
  image: /assets/images/posts/2021-01-29-web-animations-intro/og-image.png
---

[Embroider] is the future for building Ember apps. It unlocks features like
splitting code per route by statically analyzing your codebase and dependencies.
But what if you are using an addon that relies on dynamic components? This blog

> post will outline how we converted ember-promise-modals to be compatible with

Reviewer: consider linking ember-promise-modals?

Reviewer: Might also make sense to add 1 sentence quickly describing what promise-modals is?

Author: Oh yeah it was already meant to be a link, I just didn't reference it. I will also add a description.

> Embroider's routes-plitting feature.

Reviewer: ```suggestion
Embroider's route-splitting feature.
```

unless this is a joke I don't understand :)

[embroider]: https://github.com/embroider-build/embroider
[ember-promise-modals]: https://github.com/simplabs/ember-promise-modals

<!--break-->

![Making your dynamic Ember components work with Embroider today](/assets/images/posts/2021-01-29-web-animations-intro/illustration.svg#full)

## What are dynamic components?

Dynamic components are components resolved at run-time rather than hardcoding
the component to use. Ember provides a component helper which takes an argument

> that is the dasherized string representation of the component path:

Reviewer: Technically I think it's not the path but the container key that's usually based on the path (not sure that's actually relevant here and I might be wrong about it :) )?

`{{component "my-component"}}` or `{{component "folder/another-component"}}`.

> This is great for addons like ember-promise-modals as it allows us to, in this

Reviewer 2: "This is great for addons like ember-promise-modals as in this case it allows us to open a modal from javascript. "

case, open a modal from javascript.

```javascript
@service modals;

@action
confirm() {
  this.modals.open('my-component');
}
```

Internally ember-promise-modals uses the component helper to render these
modals.

## Then what is the problem?

Embroider needs to be able to statically resolve components at build time to

> provide the route-splitting feature, which is not guaranteed with the component

Reviewer: Not sure this is immediately obvious to everyone – might make sense to include a brief explanation here: you need to know which components are used so you can leave all others out of the bundle etc.

helper syntax. Hypothetically one could receive the component name from an API

> call, meaning there is no way to know this at build time.

Reviewer: Isn't the simpler example that the component name is in a variable or argument (e.g. the argument of the `open` method above) and that's already enough to make it sufficiently hard to statically analyze it so it can't really be done? Or would Embroider be able to detect that the above example is an invocation of the `<MyComponent>` component?

> Fortunately Embroider provides a few tools for us to make this work.

Reviewer 2: ```suggestion
Fortunately, Embroider provides a few tools for us to make this work.
```

## Making old addons work in your Embroider Optimized app using `packageRules`

`packageRules` are more of a compatability feature rather than being the ideal
solution. They provide a way to tell Embroider what it needs to do. The main use
case is for addons out of your control and/or addons that have not been updated

> yet to be fully Embroider optimized. By default Embroider currently ships

Reviewer 2: ```suggestion
to be fully Embroider optimized yet. By default Embroider currently ships
```

Author: I've just dropped the `yet` here  :P

`packageRules` for a few widely used addons so that they'll work out of the box.

Now let's see if we can make ember-promise-modals <= 2 work with Embroider
through `packageRules`.

If you have created an Embroider enabled app (for example through

> `ember new my-app --embroider) your `ember-cli-build.js` file will contain a

Reviewer: ```suggestion
`ember new my-app --embroider`) your `ember-cli-build.js` file will contain a
```

section that looks like this:

```javascript
const { Webpack } = require('@embroider/webpack');
return require('@embroider/compat').compatBuild(app, Webpack, {
  skipBabel: [
    {
      package: 'qunit',
    },
  ],
});
```

In order to be able to use route-splitting, we'll first have to enable all of

> Embroider's flags to ensure it will work. Normally you would do this one by one,

Reviewer 2: ```suggestion
Embroider's flags. Normally you would do this one by one,
```

but in this case the only problem we're going to run into is with the
`staticComponents` flag.

```javascript
const { Webpack } = require('@embroider/webpack');
return require('@embroider/compat').compatBuild(app, Webpack, {
  staticAddonTestSupportTrees: true,
  staticAddonTrees: true,
  staticHelpers: true,
  staticComponents: true,
  skipBabel: [
    {
      package: 'qunit',
    },
  ],
});
```

When now trying to run the app with ember-promise-modals, we'll run into a
compilation error.

```shell
Unsafe dynamic component: @modal._name in node_modules/ember-promise-modals/templates/components/modal.hbs

> ```

Reviewer: I'm not following where this is coming from tbh. Is this error coming from something like

```js
this.modals.open('epm-modal');
```

Author: not directly, it is caused by the component helper being invoked with a dynamic variable (`@modal._name`)

Let's add a `packageRules` section for the `EpmModal` component. This component
takes a `@modal` argument which is an object that also contains the `_name`
property as shown in the error that Embroider threw. We can tell Embroider that
this argument represents a component name. The layout of the component also
needs to be explicitly passed.

```javascript
const { Webpack } = require('@embroider/webpack');
return require('@embroider/compat').compatBuild(app, Webpack, {
  ...
  packageRules: [
    {
      package: 'ember-promise-modals',
      components: {
        '<EpmModal />': {
          acceptsComponentArguments: ['@modal._name'],
          layout: {
            addonPath: 'templates/components/modal.hbs',
          },
        },
      },
    },
  ],
});
```

If we now run the app Embroider will no longer throw build-time errors and our
modal will work. An unfortunate side-effect of this setup is that it will not
result in the `<ExampleModal/>` component being split from the main bundle if
you enable route splitting. In order to get that working we'll have to dig a
little deeper, but the `packageRules` approach is a good way to unblock a

> project from using a fully enabled Embroider with addons that do not yet have

Reviewer 2: ```suggestion
project from using a fully enabled Embroider with addons that do not have
```

> full Embroider support.

Reviewer 2: ```suggestion
full Embroider support yet.
```

Author: Also dropped the yet here

## Updating your addon or dynamically invoked components to be Embroider Optimized

> In order to let Embroider know how to handle our dynamic modal component we need

Reviewer 2: ```suggestion
In order to let Embroider know how to handle our dynamic modal component, we need
```

to use the `ensure-safe-component` helper that Embroider provides. This helper
will turn a component class into a component definition that can be invoked in

> the template. If just the name of a component is passed it will use the old

Reviewer 2: ```suggestion
the template. If just the name of a component is passed, it will use the old
```

> curly component resolver to get the component definition but also throw a

Reviewer 2: ```suggestion
curly component resolver to get the component definition, but also throw a
```

deprecation warning that you'll need to pass the component class when using

> Embroider. For comprehensive documentation see the

Reviewer 2: ```suggestion
Embroider. For comprehensive documentation see:
```

[Replacing the Component Helper](https://github.com/embroider-build/embroider/blob/5fd49b50dd82bf7ceb6adeefa12efc2b85f92cd2/REPLACING-COMPONENT-HELPER.md)

> documentation.

Reviewer 2: ```suggestion

```

In ember-promise-modals dynamic modal components are internally invoked with the
component helper as follows:

```handlebars
{{component @modal._name data=@modal._data close=(action 'close')}}
```

The relevant bit for us here is the first argument `@modal._name` which is the
name of the modal component, say `example-modal`. We can wrap this with the
`ensure-safe-component` helper that Embroider provides like this:

```handlebars
{{component
  (ensure-safe-component @modal._name)
  data=@modal._data
  close=(action 'close')
}}
```

Or if we want to use angle bracket syntax:

```handlebars
{{#let (component (ensure-safe-component @modal._name)) as |ModalComponent|}}
  <ModalComponent @data={{@modal._data}} @close={{action 'close'}} />
{{/let}}
```

The other thing we need to change is the way we pass the component to
ember-promise-modals in our app. We are currently still passing the
`<ExampleModal/>` component as a dynamic string.

```javascript
@service modals;

@action
confirm() {
  this.modals.open('example-modal');
}
```

> If we were to start our app now (with `staticComponents: false`) we will get the

Reviewer 2: ```suggestion
If we were to start our app now (with `staticComponents: false`), we would get the
```

> following deprecation message.

Reviewer 2: ```suggestion
following deprecation message:
```

```
DEPRECATION: You're trying to invoke the component "example-modal"
 by passing its name as a string. This won't work under Embroider.
[deprecation id: ensure-safe-component.string] See https://github.com/embroider-build/embroider/blob/master/ADDON-AUTHOR-GUIDE.md#when-youre-passing-a-component-to-someone-else for more details.
```

We can update our app code to actually import the component class so that
Embroider can statically resolve this component. This will also make the

> deprecation message go away. Note that this will _only_ work for co-located

Reviewer 2: ```suggestion
deprecation message disappear. Note that this will _only_ work for co-located
```

components or classic components that explicitly have their template definition
set on the component class using `layout`.

```javascript
import ExampleModal from '../components/example-modal';

...

@service modals;

@action
confirm() {
  this.modals.open(ExampleModal);
}
```

After re-enabling `staticComponents: true`, the last thing we need to do is
enable route splitting in our app. This can be done by modifying the `router.js`

> file to use `@embroider/router`.

Reviewer 2: ```suggestion
file to use `@embroider/router`...
```

```javascript
// app/router.js

//import EmberRouter from '@ember/routing/router';
import EmberRouter from '@embroider/router';

export default class Router extends EmberRouter {
  location = config.locationType;
  rootURL = config.rootURL;
}

Router.map(function () {});
```

> And by configuring the `splitAtRoutes` feature in `ember-cli-build.js`. We can

Reviewer 2: ```suggestion
...and by configuring the `splitAtRoutes` feature in `ember-cli-build.js`. We can
```

> do this by adding the route names you want to split or by providing a regex. Our

Reviewer 2: ```suggestion
do this by adding the route names we want to split or by providing a regex. Our
```

> full configuration will now look like:

Reviewer 2: ```suggestion
full configuration will now look like this:
```

```javascript
const { Webpack } = require('@embroider/webpack');
return require('@embroider/compat').compatBuild(app, Webpack, {
  staticAddonTestSupportTrees: true,
  staticAddonTrees: true,
  staticHelpers: true,
  staticComponents: true,
  skipBabel: [
    {
      package: 'qunit',
    },
  ],
  splitAtRoutes: ['my-page'],
});
```

> If we now start our Embroider enabled app we will see that our `<ExampleModal/>`

Reviewer 2: ```suggestion
If we now start our Embroider enabled app, we will see that our `<ExampleModal/>`
```
